Content.Server/Silicons/RobotFenceSystem.cs

```diff
@@ -0,0 +1,183 @@
+
+        var localRotation = entTransform.LocalRotation;
+
+        var startCoords = mapCoords.Position;
+        startCoords -= (localRotation + gridRotation).RotateVec(new Vector2(XOff, YOff));
+
+        // Get information ready to cast a ray
+        var mapId = entTransform.MapID;
+        // var mask = (int)CollisionGroup.HighImpassable;
+        // var mask = (int)(CollisionGroup.Opaque | CollisionGroup.HighImpassable);
+        // var mask = (int)(CollisionGroup.LowImpassable);
+        var mask = (int)(CollisionGroup.InteractImpassable);
+
+        var ray = new CollisionRay(startCoords, direction.ToVec(), mask);
+        // Cast a ray to see how far the beam goes
+        var res = _physics.IntersectRay(mapId, ray, (float)ent.Comp.BeamRange-YOff-0.5f, ent.Owner, false);
+        // If it didn't hit anything, use the maximum range
+        var distance = res.FirstOrNull()?.Distance ?? (float)ent.Comp.BeamRange;
+
+        if (distance <= 0f)
+        {
+            DestroyBeam(ent);
+            return;
+        }
+
+        var mapGridComp = Comp<MapGridComponent>(entTransform.GridUid.Value);
+
+        var beamSubSteps = (int)Math.Ceiling(distance*4f/mapGridComp.TileSize);
+
+        if (ent.Comp.BeamSubSteps == beamSubSteps)
+        {
+            return;
+        }
+
+
+
+        var curEndcapSubSteps = ent.Comp.BeamSubSteps % 4;
+        if (curEndcapSubSteps > 0)
+        {
+            QueueDel(ent.Comp.BeamEntities.Pop());
+        }
+
+        var newFullSegments = beamSubSteps / 4;
+        while (newFullSegments < ent.Comp.BeamEntities.Count)
+        {
+            QueueDel(ent.Comp.BeamEntities.Pop());
+        }
+
+        while (newFullSegments > ent.Comp.BeamEntities.Count)
+        {
+            var newCoords = new EntityCoordinates(ent.Owner, 0, -mapGridComp.TileSize*ent.Comp.BeamEntities.Count);
+            var isEmpty = newCoords.GetTileRef(EntityManager)?.Tile.IsEmpty ?? true;
+            EntityUid? newEnt = null;
+            if (!isEmpty)
+            {
+                newEnt = SpawnAtPosition("RobotFenceSegment4", newCoords);
+                _transform.SetLocalRotationNoLerp(newEnt.Value, localRotation.Opposite());
+            }
+
+            ent.Comp.BeamEntities.Add(newEnt);
+        }
+
+        var newEndcapSubSteps = beamSubSteps % 4;
+        if (newEndcapSubSteps > 0)
+        {
+            var newCoords = new EntityCoordinates(ent.Owner, 0, -mapGridComp.TileSize*ent.Comp.BeamEntities.Count);
+            var isEmpty = newCoords.GetTileRef(EntityManager)?.Tile.IsEmpty ?? true;
+            EntityUid? newEnt = null;
+            if (!isEmpty)
+            {
+                newEnt = SpawnAtPosition($"RobotFenceSegment{newEndcapSubSteps}", newCoords);
+                _transform.SetLocalRotationNoLerp(newEnt.Value, localRotation.Opposite());
+            }
+            ent.Comp.BeamEntities.Add(newEnt);
+        }
+
+        ent.Comp.BeamSubSteps = beamSubSteps;
+    }
+
```

Comment:
There's no reason to spam entities when all it needs to do is override the graph, at least until it's rewritten. Especially no reason to be checking and deleting these every single second.

The slightly less bad thing to do is just update the fixture manually to compensate but this is too expensive atm.

Reply:
I tried doing this with a single fixture, but Pathfinding.Grid.cs seems pretty insistent on throwing away fixtures that aren't centered in the current tile when updating a poly, causing some strange effects like this:

![image](https://github.com/user-attachments/assets/1f8abdbd-3ca4-486b-827d-ce2518429678)